File: BackEnd-Siukat/routes/pdf.go
package routes

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/middlewares"
	"BackEnd-Siukat/models"
	"BackEnd-Siukat/services"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
)

// PdfRoutes menangani migrasi dari routes/pdf.js Node.js
func PdfRoutes(r *gin.RouterGroup) {
	pdfGroup := r.Group("/pdf")

	// Helper untuk merender HTML sebagai penganti html-pdf
	// Di Go, tanpa dependensi *wkhtmltopdf* external, standar terbaik adalah 
	// me-render HTML murni lalu dikirim ke browser dengan trigger Print()
	// atau menggunakan library eksternal (unidoc/maroto) jika butuh re-draw murni.

	// ==========================================
	// 1. ENDPOINT STATIS (Mengunduh File PDF asli)
	// ==========================================

	pdfGroup.GET("/alur", func(c *gin.Context) {
		c.HTML(http.StatusOK, "alur.html", nil)
	})

	pdfGroup.GET("/surat-pernyataan-ukt-atas", func(c *gin.Context) {
		c.File("views/pdf/contoh/SURAT_PERNYATAAN_UKT_TERTINGGI.pdf")
	})

	pdfGroup.GET("/tidak-sanggup", func(c *gin.Context) {
		c.File("views/pdf/contoh/FORM_KETIDAKSANGGUPAN_UKT.pdf")
	})

	pdfGroup.GET("/pengumuman-snmptn", func(c *gin.Context) {
		year := time.Now().Format("2006")
		c.File(fmt.Sprintf("public/pdf/Pengumuman_Kelulusan_SNMPTN_%s.pdf", year))
	})

	pdfGroup.GET("/pengumuman-sbmptn", func(c *gin.Context) {
		year := time.Now().Format("2006")
		c.File(fmt.Sprintf("public/pdf/Pengumuman_Kelulusan_SBMPTN_%s.pdf", year))
	})

	pdfGroup.GET("/pengumuman-japres", func(c *gin.Context) {
		year := time.Now().Format("2006")
		c.File(fmt.Sprintf("public/pdf/Pengumuman_Kelulusan_JAPRES_%s.pdf", year))
	})

	pdfGroup.GET("/pengumuman-mandiri", func(c *gin.Context) {
		year := time.Now().Format("2006")
		c.File(fmt.Sprintf("public/pdf/Pengumuman_Kelulusan_MANDIRI_%s.pdf", year))
	})

	// ==========================================
	// 2. ENDPOINT DINAMIS (Butuh Autentikasi JWT)
	// ==========================================
	
	authGroup := pdfGroup.Group("/")
	authGroup.Use(middlewares.JwtAuth())

	authGroup.GET("/slip-pembayaran", func(c *gin.Context) {
		noPeserta, _ := c.Get("no_peserta")
		np := noPeserta.(string)

		// [Langkah 2] Panggil Kurir WSDL (Smart Sync)
		wsdlService := services.WsdlService{}
		if err := wsdlService.SyncToWsdl(np); err != nil {
			fmt.Printf("⚠️ WSDL SYNC WARNING: %v\n", err)
		}

		// Tarik model Info
		var info models.Info
		config.DB.First(&info)

		// Tarik data Mahasiswa dengan Preload Relasi (Gunakan Service agar Smart Priority Sanggah > Original)
		cmahasiswaService := services.CMahasiswaService{}
		mhs, err := cmahasiswaService.GetCmahasiswa(np)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Data Mahasiswa tidak ditemukan"})
			return
		}
		
		// Pastikan relasi Prodi & Fakultas ke-load
		config.DB.Preload("Fakultas").Preload("Prodi").Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").First(&mhs, mhs.IDCmahasiswa)

		// Tarik UKT berdasarkan Golongan & Mayor
		var ukt models.Ukt
		config.DB.Where("major_id = ?", mhs.ProdiCmahasiswa).First(&ukt)

		// Karena kita mendelegasikan PDF creation ke Golang HTML Templates:
		// Variabel data (ekuivalen obj di NodeJS)
		data := gin.H{
			"info":       info,
			"cmahasiswa": mhs,
			"ukt":        ukt,
		}

		// RENDER: Panggil Go Template "slip-pembayaran.html" 
		// *Pastikan file ejs lama sudah diubah sintaks <% %> -nya menjadi {{ }} ala Go Templates*
		c.HTML(http.StatusOK, "slip-pembayaran.html", data)
	})

	authGroup.GET("/bukti-selesai", func(c *gin.Context) {
		noPeserta, _ := c.Get("no_peserta")
		np := noPeserta.(string)

		var info models.Info
		config.DB.First(&info)

		var mhs models.CMahasiswa
		if err := config.DB.Preload("Fakultas").Preload("Prodi").Where("no_peserta = ?", np).First(&mhs).Error; err != nil {
			c.JSON(http.StatusInternalServerError, err.Error())
			return
		}

		data := gin.H{
			"info":       info,
			"cmahasiswa": mhs,
		}

		c.HTML(http.StatusOK, "bukti-selesai.html", data)
	})

	authGroup.GET("/surat-validasi", func(c *gin.Context) {
		noPeserta, _ := c.Get("no_peserta")
		np := noPeserta.(string)

		// Tarik seluruh kompenen
		var mhs models.CMahasiswa
		var a models.Ayah
		var i models.Ibu
		var k models.Kendaraan
		var w models.Wali
		var r models.Rumah
		var l models.Listrik
		var p models.Pendukung

		config.DB.Preload("Fakultas").Preload("Prodi").Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").Where("no_peserta = ?", np).First(&mhs)
		config.DB.Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").Preload("Pekerjaan").Where("no_peserta = ?", np).First(&a)
		config.DB.Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").Preload("Pekerjaan").Where("no_peserta = ?", np).First(&i)
		config.DB.Where("no_peserta = ?", np).First(&k)
		config.DB.Where("no_peserta = ?", np).First(&w)
		config.DB.Where("no_peserta = ?", np).First(&r)
		config.DB.Where("no_peserta = ?", np).First(&l)
		config.DB.Where("no_peserta = ?", np).First(&p)

		data := gin.H{
			"tahun":      time.Now().Format("2006"),
			"cmahasiswa": mhs,
			"ayah":       a,
			"ibu":        i,
			"kendaraan":  k,
			"wali":       w,
			"rumah":      r,
			"listrik":    l,
			"pendukung":  p,
		}

		c.HTML(http.StatusOK, "surat-validasi.html", data)
	})

	authGroup.GET("/kontrak", func(c *gin.Context) {
		noPeserta, _ := c.Get("no_peserta")
		np := noPeserta.(string)

		var mhs models.CMahasiswa
		config.DB.Preload("Fakultas").Preload("Prodi").Where("no_peserta = ?", np).First(&mhs)

		data := gin.H{
			"cmahasiswa": mhs,
		}
		c.HTML(http.StatusOK, "kontrak.html", data)
	})

	authGroup.GET("/wali", func(c *gin.Context) {
		noPeserta, _ := c.Get("no_peserta")
		np := noPeserta.(string)

		var mhs models.CMahasiswa
		var w models.Wali

		config.DB.Preload("Fakultas").Preload("Prodi").Where("no_peserta = ?", np).First(&mhs)
		config.DB.Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").Where("no_peserta = ?", np).First(&w)

		data := gin.H{
			"cmahasiswa": mhs,
			"wali":       w,
		}
		c.HTML(http.StatusOK, "wali.html", data)
	})

	authGroup.GET("/sanggah", func(c *gin.Context) {
		// Logika ini diperuntukkan narik Array of Sanggah Data
		var mhsList []models.CMahasiswa
		config.DB.Preload("Fakultas").Preload("Prodi").Where("flag = ?", "sanggah_ukt").Find(&mhsList)

		data := gin.H{
			"rows": mhsList,
		}
		c.HTML(http.StatusOK, "pdf-sanggah.html", data)
	})

	authGroup.GET("/bm", func(c *gin.Context) {
		var mhsList []models.CMahasiswa
		config.DB.Preload("Fakultas").Preload("Prodi").Where("bidik_misi_cmahasiswa = ?", "Ya").Find(&mhsList)

		data := gin.H{
			"rows": mhsList,
		}
		c.HTML(http.StatusOK, "pdf-bm.html", data)
	})

	authGroup.GET("/master", func(c *gin.Context) {
    var mhsList []models.CMahasiswa
    cmahasiswaService := services.CMahasiswaService{}

    // 1. QUERY ANTI-DOUBLE (Smart Prioritas)
    config.DB.Model(&models.CMahasiswa{}).
        Select("tb_cmahasiswa.*"). // Pastikan ambil kolom utama
        Preload("Fakultas").Preload("Prodi").
        Joins("LEFT JOIN tb_cmahasiswa AS t2 ON tb_cmahasiswa.no_peserta = t2.no_peserta AND t2.atribut = 'sanggah'").
        Where("tb_cmahasiswa.atribut = 'sanggah' OR (tb_cmahasiswa.atribut = 'original' AND t2.id_cmahasiswa IS NULL)").
        Order("tb_cmahasiswa.no_peserta ASC").
        Find(&mhsList)

    // 2. LOOP DUIT: Hybrid Display (Nominal=Original, Tagihan=Current/Sanggah)
    for i := range mhsList {
        // A. Ambil data ORIGINAL buat penentu Nominal
        var mhsOriginal models.CMahasiswa
        config.DB.Where("no_peserta = ? AND atribut = ?", mhsList[i].NoPeserta, "original").First(&mhsOriginal)

        // B. Kolom NOMINAL UKT: Paksa nampilin Nominal Original
        var uktMaster models.Ukt
        if err := config.DB.Where("major_id = ? AND CAST(entrance AS CHAR) = ?", mhsList[i].ProdiCmahasiswa, mhsList[i].JalurCmahasiswa).First(&uktMaster).Error; err == nil {
            mhsList[i].Ukt = &uktMaster
            // Hitung pake GolonganID dari data Original
            mhsList[i].Ukt.Nominal = cmahasiswaService.CalculateNominalValue(uktMaster, mhsOriginal.GolonganID)
        }

        // C. Kolom TAGIHAN UKT: Pake data SEKARANG (Sanggah jika ada, Ori jika tidak)
        // Kita ambil dari mhsList[i].Tagihan langsung karena query sudah memprioritaskan Sanggah
        if tagihanInt, errStr := strconv.Atoi(mhsList[i].Tagihan); errStr == nil {
            mhsList[i].OriginalNominal = tagihanInt
        } else {
            mhsList[i].OriginalNominal = 0
        }
    }

    data := gin.H{
        "now":        time.Now().Format("02 January 2006, 15:04 WIB"),
        "cmahasiswa": mhsList,
    }
    c.HTML(http.StatusOK, "pdf-master.html", data)
})

}

File: ZhammDataBeLama/siukat-service-dev_snmptn_2022/routes/pdf.js
var express = require('express');
var router = express.Router();
var db = require('../config/database');
var path = require('path');
var passport = require('passport');
var { cmahasiswa, ayah, ibu, rumah, listrik, wali, kendaraan, pendukung, users } = require('../services')
var moment = require('moment')
var CMahasiswa = require('../models/cmahasiswa');
var Info = require('../models/info')
var Ukt = require('../models/ukt')
var global = require('../global')

var fs = require('fs');
var ejs = require('ejs');
var pdf = require('html-pdf')

router.get('/slip-kocak', function (req, res, next) {
  // Check Tagihan dulu
  const tagihan = "9219013497";
  cmahasiswa.cekTagihan(tagihan).then(rs => {
    // Generate PDF nya
    var obj = {}
    Info.findOne().then(function (response) {
      response.pembayaran = "14 Mei 2019 - 16 Agustus 2019"
      obj.info = response
      return users.getUser(tagihan)
    }).then(function (response) {
      obj.cmahasiswa = response
      return Ukt.findOne({
        where: {
          major_id: response.prodi_cmahasiswa
        },
        attributes: [response.golongan_id]
      })
    }).then(function (response) {
      obj.ukt = response
      var compiled = ejs.compile(fs.readFileSync('views/pdf/slip-pembayaran.html', 'utf8'))
      var html = compiled(obj)
      var options = { format: 'A4' }
      pdf.create(html, options).toStream(function (err, stream) {
        stream.pipe(res);
      });
    }).catch(function (err) {
      res.status(500).json(err)
    })
  }).catch(err => {
    res.status(500).json(err)
  })
})

router.get('/slip-pembayaran', passport.authenticate('jwt', { session: false }), function (req, res, next) {
  // Check Tagihan dulu
  cmahasiswa.cekTagihan(req.user.no_peserta).then(rs => {
    // Generate PDF nya
    var obj = {}
    Info.findOne().then(function (response) {
      obj.info = response
      return users.getUser(req.user.no_peserta)
    }).then(function (response) {
      obj.cmahasiswa = response
      return Ukt.findOne({
        where: {
          major_id: response.prodi_cmahasiswa
        },
        attributes: [response.golongan_id]
      })
    }).then(function (response) {
      obj.ukt = response
      var compiled = ejs.compile(fs.readFileSync('views/pdf/slip-pembayaran.html', 'utf8'))
      var html = compiled(obj)
      var options = { format: 'A4' }
      pdf.create(html, options).toStream(function (err, stream) {
        stream.pipe(res);
      });
    }).catch(function (err) {
      res.status(500).json(err)
    })
  }).catch(err => {
    res.status(500).json(err)
  })
})

router.get('/bukti-selesai', passport.authenticate('jwt', { session: false }), function (req, res, next) {
  var obj = {}
  Info.findOne().then(function (response) {
    obj.info = response
    return users.getUser(req.user.no_peserta)
  }).then(function (response) {
    obj.cmahasiswa = response
    var compiled = ejs.compile(fs.readFileSync('views/pdf/bukti-selesai.html', 'utf8'))
    var html = compiled(obj)
    var options = { format: 'A4', base: "file:///Applications/XAMPP/htdocs/siukat-service/public/" }
    pdf.create(html, options).toStream(function (err, stream) {
      stream.pipe(res);
    });
    // .toFile('./result.pdf',() => {
    //     res.pdf('./result.pdf')
    // })
  }).catch(function (err) {
    res.json(err)
  })
})

router.get('/surat-validasi', passport.authenticate('jwt', { session: false }), function (req, res, next) {
  var obj = {
    tahun: moment().format('YYYY')
  }

  users.getUser(req.user.no_peserta).then((response) => {
    obj.cmahasiswa = response
    return ayah.getByLoggedIn(req.user.no_peserta)
  }).then((response) => {
    obj.ayah = response
    return ibu.getByLoggedIn(req.user.no_peserta)
  }).then((response) => {
    obj.ibu = response
    return kendaraan.getByLoggedIn(req.user.no_peserta)
  }).then((response) => {
    obj.kendaraan = response
    return wali.getByLoggedIn(req.user.no_peserta)
  }).then((response) => {
    obj.wali = response
    return rumah.getByLoggedIn(req.user.no_peserta)
  }).then((response) => {
    obj.rumah = response
    return listrik.getByLoggedIn(req.user.no_peserta)
  }).then((response) => {
    obj.listrik = response
    return pendukung.getByLoggedIn(req.user.no_peserta)
  }).then((response) => {
    obj.pendukung = response
    var compiled = ejs.compile(fs.readFileSync('views/pdf/surat-validasi.html', 'utf8'))
    var html = compiled(obj)
    var options = {
      format: 'A4',
      base: "file:///Applications/XAMPP/htdocs/siukat-service/public/",
    }
    pdf.create(html, options).toStream(function (err, stream) {
      stream.pipe(res);
    });
  }).catch(function (err) {
    res.json(err)
  })
})

router.get('/sanggah', passport.authenticate('jwt', { session: false }), function (req, res, next) {
  cmahasiswa.getSanggah().then(function (response) {
    var response = {
      rows: response
    }

    var compiled = ejs.compile(fs.readFileSync('views/pdf/pdf-sanggah.html', 'utf8'))
    var html = compiled(response, options)
    var options = { format: 'A4', orientation: 'landscape', base: "file:///F:/xampp/htdocs/siukat-service/public/" }
    pdf.create(html, options).toStream(function (err, stream) {
      stream.pipe(res);
    });
  }).catch(function (err) {
    res.json(err)
  })
})

router.get('/bm', passport.authenticate('jwt', { session: false }), function (req, res, next) {
  cmahasiswa.getBM().then(function (response) {
    response = {
      rows: response
    }

    var compiled = ejs.compile(fs.readFileSync('views/pdf/pdf-bm.html', 'utf8'))
    var html = compiled(response)
    var options = { format: 'A4', orientation: 'landscape', base: "file:///F:/xampp/htdocs/siukat-service/public/" }
    console.log(html)
    pdf.create(html, options).toStream(function (err, stream) {
      stream.pipe(res);
    });
  }).catch(function (err) {
    res.json(err)
  })
})

router.get('/kontrak', function (req, res, next) {
  var obj = {
    tahun: moment().format('YYYY')
  }

  var compiled = ejs.compile(fs.readFileSync('views/pdf/kontrak.html', 'utf8'))
  var html = compiled(obj)

  pdf.create(html).toStream(function (err, stream) {
    stream.pipe(res);
  });

  /*pdf.create(html).toBuffer(function(err, buffer){
    res.send(buffer)
  });*/
})

router.get('/wali', function (req, res, next) {
  res.pdf(path.resolve('views/pdf/contoh/SURAT_PERNYATAAN_WALI_CALON_MAHASISWA.pdf'))
})

router.get('/alur', function (req, res, next) {
  var obj = {
    tahun: moment().format('YYYY')
  }

  var compiled = ejs.compile(fs.readFileSync('views/pdf/alur.html', 'utf8'))
  var html = compiled(obj)
  pdf.create(html).toStream(function (err, stream) {
    stream.pipe(res);
  });
})

router.get('/surat-pernyataan-ukt-atas', function (req, res, next) {
  res.pdf(path.resolve('views/pdf/contoh/SURAT_PERNYATAAN_UKT_TERTINGGI.pdf'))
})

router.get('/tidak-sanggup', function (req, res, next) {
  res.pdf(path.resolve('views/pdf/contoh/FORM_KETIDAKSANGGUPAN_UKT.pdf'))
})

router.get('/pengumuman-snmptn', (req, res, next) => {
  const year = new Date().getFullYear();
  res.pdf(path.resolve('public/pdf/Pengumuman_Kelulusan_SNMPTN_' + year + '.pdf'))
})

router.get('/pengumuman-sbmptn', (req, res, next) => {
  const year = new Date().getFullYear();
  res.pdf(path.resolve('public/pdf/Pengumuman_Kelulusan_SBMPTN_' + year + '.pdf'))
})

router.get('/pengumuman-japres', (req, res, next) => {
  const year = new Date().getFullYear();
  res.pdf(path.resolve('public/pdf/Pengumuman_Kelulusan_JAPRES_' + year + '.pdf'))
})

router.get('/pengumuman-mandiri', (req, res, next) => {
  const year = new Date().getFullYear();
  res.pdf(path.resolve('public/pdf/Pengumuman_Kelulusan_MANDIRI_' + year + '.pdf'))
})

module.exports = router;
